### ISSUE TITLE
Set up gitlab continuous integration (CI)

### ISSUE THREAD

**Issue by Blebowski** [Logic-Design-Services/CTU-CAN-FD#125]
***In GitLab by @mjerabek on Apr 16, 2018, 18:41 GMT+2:***

Set up basic CI/CD to run short sanity test after each commit. For now, use the available shared runner.

**Assignees:** @mjerabek

*Migrated from GitLab: https://gitlab.fel.cvut.cz/canbus/ctucanfd_ip_core/-/issues/122*

**Comment by Blebowski** [comment]
***In GitLab by @mjerabek on May 21, 2018, 12:38 GMT+2:***

created branch [`122-set-up-gitlab-continuous-integration-ci`](https://gitlab.fel.cvut.cz/illeondr/CAN_FD_IP_Core/compare/master...122-set-up-gitlab-continuous-integration-ci)

**Comment by Blebowski** [comment]
***In GitLab by @mjerabek on May 21, 2018, 21:55 GMT+2:***

So, the current setup is following:
- after each push, build everything (src+unit+sanity) and run all unit tests
- every night run sanity test on master

If anything fails, we apparently get an email.

Feature tests are not run at the moment, because GHDL does not yet support VHDL-2008 external signals. I guess after 10 years it's still new :/ 

The checking happens after every push (if there is `.gitlab-ci.yml` present), which may be
- a) needlessly taxing on the runner if there are no changes in code (e.g. driver commits etc.)
- b) too noisy when on development topic branches

We might run the tests only on the master branch, but then we would have to run them on merge requests manually (or find a hack), because gitlab is rather inflexible here. For now, I'd just wait and see how much annoying it will be.

**Comment by Blebowski** [comment]
Hi,

great work :owl: :eyes: :thumbsup\_tone3: 

Some ideas though:

1. Protocol control unit test will be updated, OK to move it to "." prefix folder for now. I will move it back once fixed. Let's make it rule that if an update is made which fails a test, it will be moved to directory with ".". I didn't check, but I assume you filter directories for test selection based on "." in name or first character?

2. Every push, no problem, I will have to re-evaluate my push policy since I usually push after each commit. Having runs just due to change "dummy" change in everything might not be good...

3. External signals from VHDL 2008 in feature tests is a puzzle. Do I understand properly that the 
   problem is in usage of "alias" for access to internal signals inside Feature environment wrapper??
   I wanted to have posibility to look at internal signals (in case something not in register map
   will be needed in test logic). I did use it in few spots only. E.g. "stat_bus" in 
   "spec_mode_feature" where I polled on Protocol control state.
   Option would be to bring out all the signals from "CAN_feature_test" entity, but then the same
   problem would be one level lower... And I really don't want to bring out status bus out of 
   CAN_top_level.vhd. Options are:
      A) Bring out only signals which are in "CAN_feature_test" and not internal to CAN Core. Those
         can be just added to ports. Remove internal signals. Modify test logic not to check any
         internal signals.
      B) Well there are no other options, unless we want to write VHDL 2008 support in GHDL...

4. Most important. I looked up test runs of unit test. All of the tests take 0.5 seconds which is
   way too little. More importantly all the tests are different and run for significantly different 
   amount of time (with the same "iterations"). Are you sure all tests were truly executed in GHDL?
   I would suggest setting "log_level" to "info_l" and checking the output. Each iteration should
   print output to test log with "Note" severity.

**Comment by Blebowski** [comment]
***In GitLab by @mjerabek on May 22, 2018, 01:00 GMT+2:***

Hi,

1. The "*" in glob search pattern does ignore hidden files, I was just lazy and used this :) I also didn't want to move it too far away so that it is not forgotten. On the other hand, when we make an update which breaks a test, we should either fix the source or the test, not just remove the test :D (but I understand and agree).

2. That is what I don't like. Frequent pushes are good, but frequent checks are unneeded. Maybe really just enable it for master and manually before merge? I'll set it up this way tomorrow and we'll see if it is enough.

3. The problem with GHDL is everything. I remember trying every possible variant (alias, direct use) with external signals, yet alas all in vain. It always crashed during compilation and I had to resort to pulling the signals from the innards of the component. With only CAN_feature_test it sounds feasible, but the rest ... I don't like removing the checks of the internals, we would be sacrificing bug-catching potential. And pulling all the signals out ... maybe through a middle layer or some configuration magic, but I still don't like it.
I think there could be option B: VUnit is simulator-agnostic, so we could just run it using modelsim. But:

  - we wouldn't have coverage for this (but still the tests)
  - we would have to create docker image with modelsim; that would have to be private (licenses), which is possible (docker hub offers 1 private project in the free plan), but then I don't know how to make it visible to gitlab.

I'll try to think about it.
If GHDL wasn't written in Ada, I might try to poke in it, but ...

.4. That's a very good point. I was happy that all the hacks work that I didn't notice they probably don't. I'll investigate.

Thanks for the feedback.

**Comment by Blebowski** [comment]
Hi,

I didn't notice new run in pipeline after "dummy" Merge request. I will not merge it right now since I am not sure whether you managed to bring "execute after merge request" up.

I will progress and merge other stuff, no problem to rebase the branch later on.

Btw. thanks for remote access, I got to the PC, and APO board. Checked device tree. I did not run Vivado
yet, neither rebooted the board, neither tried the driver. I first would like to run GHDL run with coverage.
APO board, TFTP boot, GHDL, VUnit are new toolchain for me so it takes me some time understand
everything. I rather would not like to hastly experiment with the boot and crash the board there.
